docs/proposed/pub-sub.rst

```diff
@@ -0,0 +1,203 @@
+     can obviously "invite" further clients
+   - nobody else may subscribe
+ - to add information, the publisher adds a new file to the mutable
+   directory
+ - old information can be removed (or not) as the publisher wishes
+
+Without further changes to Tahoe-LAFS, this can already be achieved:
+clients can "poll" the read-capability and determine if anything new
+exists.
+
+
+Improvements Over Polling
+-------------------------
+
+The above mechanism would scale somewhat poorly as the number of
+"broadcast" capabilities increased. It would be better if the
+storage-server could *tell* clients when a mutable Slot has changed.
+
+Storage servers implement a new feature, based on WebSockets. See also
+:ref:`http-storage-node-protocol` with which this protocol aims to be
+compatible. A single WebSocket endpoint exists:
+
+ - <server>/v1/mutable_updates
+
+After connecting to this endpoint, a client may send any number of
+messages (encoded using JSON) asking for updates to mutable
+Slots:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "subscribe": [
+            "storage-index-0",
+            "storage-index-1",
+        }
+    }
+
+..where "storage-index-0" corresponds to an actual Storage Index of an
+existing mutable file. A client computes this from "R" (a read
+capability). The client must keep a mapping of `Storage Index ->
+read-capability` so it can match subsequent notifications.
+
+The server replies with a message like:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "status": {
+            "storage-index-0": true,
+            "storage-index-1": "some kind of error message"
+        }
+    }
+
+That is, a "status" dict containing each requested Slot's
+storage-index with either "true" if the subscribe was successful or a
+string with an error-message if it was not possible to subscribe
+(e.g. the capability wasn't a mutable one, or wasn't found on this
+server). Every storage-index requested in the initial list will have
+an entry in this dict. The client should remember a map of all
+*successful* subscriptions back to the corresponding read-capability.
+
+TODO:
+
+ - for a given Slot, does a client just ask every storage-server for
+   the corresponding Storage Index? (are there information-disclosure
+   issues here?)
+
+       meejah: a storage-server can learn which Slots a client is
+       interested in, but it learns this information with the "polling"
+       mechanism too -- so i think the answer is, "to storage-server
+       learns the same information as with the polling method"
+
+ - do we want a more-structured response for errors, e.g. with
+   machine-readable "code" and a human-readable "message"?
+
+All subscriptions for a particular client shall be removed when that
+client's WebSocket connection goes away.
+
+The server will send a single message for each update:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "update": "storage-index-1"
```

Comment:
Maybe this should be `"updates": [...]` and storage servers should be allowed to delay notification briefly to possibly bundle up several changes in one notification.

Would this actually be a win?  I suppose for a highly active magic-folder you might sometimes see two different peers changing their own version of the files at close to the same time.  You're going to have to watch each of their DMDs so it might be nice to get one update that says N clients have changed files.  Eh.  It sounds pretty thin.  Possibly not worth it.

Reply:
As a further thought to that, perhaps the subscribing client could include an option `maximum-delay` or similar specifying how many seconds the server can wait before sending updates.

Thinking about this, there's huge tension trying to make a "generic" protocol here I feel .. if you're doing real-time collaborative editing, you don't want to wait "at all". In that case, perhaps a direct peer-to-peer style connection is better? On the other hand, if you're NOT doing real-time editing, the delay could be quite a bit longer. All that said, the proposal here does seem more about "real-time" edits since it depends on an active WebSocket connection .. so maybe `maximum-delay` should be in milliseconds.

Additionally, maybe we should make this more explicit in this spec: there is a mechanism to tell your client to go into "rapid update" mode (and to turn that mode off). When it's on, the client makes active WebSocket connections to the server; when it's off it does periodic polling (or even "nothing, until told otherwise"). Although we don't have a phone application, this would map well to those use-cases (as far as I can understand them): if the phone sleeps (or the app goes out of focus) we're in "polling" or even "don't care" mode; if the app is active, we want immediate updates (because presumably a user is touching their fondle-slab at that moment).

Reply:
I'm somewhat skeptical that we're going to achieve a reasonably performing real-time collaborative editing system on top of storage-server-mediated updates.  Particularly considering there are still essentially no facilities for multi-writer...  I think instead what we should be aiming for now is a "pretty fast" sync experience.

Reply:
Fair.

Still, I could imagine GridSync being smart about "my human is here" (so make WebSocket connections etc) versus "my human is gone" (e.g. screensaver turned on) so I don't really care about updates now.

Reply:
I pushed changes making that "updates"

.. and while I agree that magic-folder isn't going to be a multi-user collaborative thing in the immediate future, it's still worth thinking about these issues, especially in the context of "pub-sub" -- a CRDT (for example) could be built on top of mutables + notifications for "actual" real-time collaboration (or any other data-structure that could be stored in a mutable, which is basically "any data-structure you can serialize").